trestle_fedramp/core/ssp_reader.py

```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2024 IBM Corp. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Read and prepare OSCAL SSP information for template.
+
+The OSCAL SSP is stored in the FedRAMPControlDict in a way that
+would be expected by the FedRAMP Template.
+Control ID -> Labels
+Control Origination Property -> Control Origination String Value
+"""
+
+import pathlib
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from trestle.common.const import CONTROL_ORIGINATION, NAMESPACE_FEDRAMP
+from trestle.common.list_utils import as_list
+from trestle.common.load_validate import load_validate_model_path
+from trestle.core.catalog.catalog_interface import CatalogInterface
+from trestle.core.control_interface import ControlInterface
+from trestle.core.profile_resolver import ProfileResolver
+from trestle.oscal.catalog import Catalog
+from trestle.oscal.ssp import ImplementedRequirement, SystemSecurityPlan
+
+from trestle_fedramp.const import (
+    FEDRAMP_CUST_CONFIGURED,
+    FEDRAMP_CUST_PROVIDED,
+    FEDRAMP_HYBRID,
+    FEDRAMP_INHERITED,
+    FEDRAMP_SHARED,
+    FEDRAMP_SP_CORPORATE,
+    FEDRAMP_SP_SYSTEM
+)
+
+
+class ControlOrigination:
+    """Represents the FedRAMP control origination mapping per FedRAMP validation rules."""
+
+    SP_CORPORATE = 'sp-corporate'
+    SP_SYSTEM = 'sp-system'
+    CUST_CONFIGURED = 'customer-configured'
+    CUST_PROVIDED = 'customer-provided'
+    INHERITED = 'inherited'
+
```

Comment:
Should these short names be in a const file? Also are these defined by FedRAMP or us?

Reply:
These are defined by FedRAMP [here](https://github.com/GSA/fedramp-automation/blob/45479230b1fa712b5a8443a634593fcef1a14e44/src/content/rev5/resources/xml/fedramp_values.xml). I went back and forth over whether I we should define them here or parse the xml file. I ended up defining them since the values seem to be very stable, but I can also see how that would not scale as easily as we add more FedRAMP specific values. Interested in your thoughts since `fedramp` resources are already being copied over.

Reply:
Also wanted to add that the control origination values in the xml do not cover the `hybrid` or `shared` values that are check boxes on the template, but the FedRAMP SSP template shows examples with multiple control origination values. So the logic on handling multiple control origination values and how they translate to `hybrid` and `shared` is defined here.